ferrocene/doc/requirements/src/self-test.rst

```diff
@@ -0,0 +1,209 @@
+
+- The executable should exit successfully.
+
+- The output of the executable should be UTF-8 text.
+
+- The output of the executable should have at least 3 values: host, commit-hash, and release.
+
+- The value of ``host`` from the output of the executable should match the target-triple
+  of the executing environment.
+
+- The value of ``commit-hash`` from the output of the executable should match the 
+  commit hash of the Ferrocene release corresponding to the installed Ferrocene toolchain.
+
+- The value of ``release`` from the output of the executable should match the
+  version of the Ferrocene release corresponding to the installed Ferrocene toolchain.
+
+Checks for targets
+------------------
+
+.. list-table::
+   :header-rows: 1
+
+   * - Target
+     - Has a standard library (``std``)
+   * - :target-with-triple:`x86_64-unknown-linux-gnu`
+     - ✔
+   * - :target-with-triple:`aarch64-unknown-none`
+     -
+   * - :target-with-triple:`x86_64-pc-windows-msvc`
+     - ✔
+   * - :target-with-triple:`aarch64-unknown-linux-gnu`
+     - ✔
+   * - :target-with-triple:`aarch64-apple-darwin`
+     - ✔
+   * - :target-with-triple:`x86_64-apple-darwin`
+     - ✔
+   * - :target-with-triple:`aarch64-unknown-nto-qnx710`
+     - ✔
+   * - :target-with-triple:`x86_64-pc-nto-qnx710`
+     - ✔
+   * - :target-with-triple:`armv8r-none-eabihf`
+     -
+   * - :target-with-triple:`thumbv7em-none-eabi`
+     -
+   * - :target-with-triple:`thumbv7em-none-eabihf`
+     -
+   * - :target-with-triple:`armv7r-none-eabihf`
+     -
+   * - :target-with-triple:`armebv7r-none-eabihf`
+     -
+
+.. note::
+
+   In the subsections below, we have these substitutions:
+
+   - ``$target`` refers to any of the targets listed
+   - ``$hash`` is a 16 character hexadecimal string
+
+All targets
+^^^^^^^^^^^
+
+- Inside of ``lib/rustlib/$target/lib`` directory,
+  relative to the root of the installation,
+  should exist these regular files:
+
+  - ``libcore-$hash.rlib``
+  - ``liballoc-$hash.rlib``
+
+- Inside of ``lib/rustlib/$target/lib`` directory,
+  relative to the root of the installation,
+  should not exist duplicates of these regular files:
+
+  - ``libcore-$hash.rlib``
+  - ``liballoc-$hash.rlib``
+
+Targets with ``std``
+^^^^^^^^^^^^^^^^^^^^
+
+- Inside of ``lib/rustlib/$target/lib`` directory,
+  relative to the root of the installation,
+  should exist these regular files:
+
+  - ``libstd-$hash.rlib``
+  - ``libtest-$hash.rlib``
+  - ``libproc_macro-$hash.rlib``
+
+- Inside of ``lib/rustlib/$target/lib`` directory,
+  relative to the root of the installation,
+  should not exist duplicates of these regular files:
+    
+  - ``libstd-$hash.rlib``
+  - ``libtest-$hash.rlib``
+  - ``libproc_macro-$hash.rlib``
+
+Checks for linkers
+------------------
+
+.. note:: The following checks apply to these targets:
+
+   - :target:`x86_64-unknown-linux-gnu`
+
+   ``$target`` refers to any of the targets listed above
```

Comment:
Hmm? We check every host tool, not just linux.

Reply:
b3fa541a295